### PR TITLE
01 fib

### LINKED ISSUE #221: Lab01 Fibonacci
**Lab 1 - Fibonacci sequence**
- [ ] Create an executable go program in directory 01_fib
- [ ] Write a function that prints the first n Fibonacci numbers

### PR DESCRIPTION
**Pull request by jonathanheaden**
Fixes #221

#### Changes proposed in this PR:
Implement Fib func
Add tests for fib(7) and fib(-7)

### CODE REVIEW

File: 01_fib/jonathanheaden/main.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+)
+
+var out io.Writer = os.Stdout
+var fibs = map[int]int{}
```
Comment (camscale) [inline 01_fib/jonathanheaden/main.go]: a map where each index is a sequential number from 0 is really the same data structure as an array/slice, just less efficient.

Also do you really need to store *all* the fibs, or just the last two? Can you get away with out an indexed data structure at all?

File: 01_fib/jonathanheaden/main.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+	prefix := ""
+	var res string
+	if n < 0 {
+		n *= -1
+		prefix = "-"
+	}
+	fibs[0] = 1
+	fibs[1] = 1
+	for i := 0; i < n; i++ {
+		if i > 1 {
```
Comment (camh-anz) [inline 01_fib/jonathanheaden/main.go]: Rather than have this if statement, can you just start the loop at 2?

Comment (camh-anz) [reply 01_fib/jonathanheaden/main.go]: maybe not, since you need to print the earlier numbers...

File: 01_fib/jonathanheaden/main_test.go [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+	// Given
+	r := require.New(t)
+	var buf bytes.Buffer
+	out = &buf
+
+	// When
+	main()
+
+	// Then
+	expected := strconv.Quote("\n1\n1\n2\n3\n5\n8\n13\n\n-1\n-1\n-2\n-3\n-5\n-8\n-13\n")
```
Comment (camscale) [inline 01_fib/jonathanheaden/main_test.go]: The output of the negative fib series is not correct. Double-check the wikipedia page. The sign alternates.

File: 01_fib/jonathanheaden/main.go (hunk 1 above)
Comment (jonathanheaden) [reply 01_fib/jonathanheaden/main.go]: agree. I have replaced the map with a three element array

File: 01_fib/jonathanheaden/main_test.go (hunk 3 above)
Comment (jonathanheaden) [reply 01_fib/jonathanheaden/main_test.go]: I had to think through this a lot, I see what you mean. I have altered the test and code

**Review by juliaogris — DISMISSED** [review]
Please have a read through the [collected lab 1 review comments](https://github.com/anz-bank/go-course/wiki/Lab-1:-Collected-Code-Review-Comments):
- Create issue and add link to PR description
- Fix up markdown in Pr
- Consider using table driven tests
- Fold your 8 commits into one or two with `git rebase`

Your solution could be simplified to something like:

```go
	sign := // 1 for positive n ; -1 for negative n
	n1, n2 := 0, 1
	for j := 0; j < n*sign; j++ {
		n1, n2 = n2, n1+sign*n2
		// print n1
	}
```

File: 01_fib/jonathanheaden/main.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+	var neg = false
+	if n < 0 {
+		n *= -1
+		neg = true
+	}
+	for i := 0; i <= n; i++ {
+		if prefix = ""; neg && (i%2 == 0) {
+			prefix = "-"
+		}
+		switch i {
```
Comment (juliaogris) [inline 01_fib/jonathanheaden/main.go:27]: No need to have a switch on loop index inside look - might as well do it before imo.

**Review by camscale — APPROVED** [review]
I happy with the Go code in this PR. You could implement the loop more simply along the lines of Julia's earlier suggestion, and a `fib(1)` test case would be nice to see as well, since you have a `case 1` in the code, but these are more software engineering issues and we want to focus on Go coding for these lab PRs.

**Review by juliaogris — APPROVED** [review]
The lab is solved!

You have done everything that's required.
I still think when you calculate fib you care about the right values as integer and not just them being printed correctly, so the prefix is a bid odd to me.
To me a solution like the listing below would have been a bit cleaner:
```go
func fib(n int) {
	if n == 0 {
		return
	}

	sign := 1
	if n < 0 {
		sign = -1
	}

	n1, n2 := 0, 1
	for j := 0; j < n*sign; j++ {
		n1, n2 = n2, n1+sign*n2
		fmt.Fprintln(out, n1)
	}
}
```

**Pull request merged**